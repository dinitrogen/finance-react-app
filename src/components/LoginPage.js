import { React, useEffect, useState } from "react";
import { StyledButton, StyledInput } from "./StyledComponents";
import { auth, db } from "../firebase";
import { 
    AuthErrorCodes,
    connectAuthEmulator,
    signInWithEmailAndPassword,
    createUserWithEmailAndPassword,
    onAuthStateChanged,
    signOut,
    updateProfile
} from "firebase/auth";
import { updateDoc, doc, getDoc, arrayUnion, setDoc } from "firebase/firestore";


// connectAuthEmulator(auth, "http://localhost:9099");


const LoginPage = () => {
    
    const monitorAuthState = async () => {
        onAuthStateChanged(auth, user => {
            if (user) {
                setUserSignedIn(true);
                setDisplayName(user.displayName);
            } else {
                setUserSignedIn(false);
            }
        });
    }

    const loginEmailPassword = async (event) => {
        event.preventDefault();
        const loginEmail = userEmail;
        const loginPassword = userPassword;
        
        try {
            const userCredential = await signInWithEmailAndPassword(auth, loginEmail, loginPassword);
            setUserName('');
            setUserEmail('');
            setUserPassword('');
            setErrorMessage('');
            setDisplayName(userCredential.user.displayName);

        } catch(error) {
            console.log(error);
            showLoginError(error);
        }
    }

    const createAccount = async (event) => {
        event.preventDefault();
        const loginEmail = userEmail;
        const loginPassword = userPassword;

        try {
            const userCredential = await createUserWithEmailAndPassword(auth, loginEmail, loginPassword);
            // await updateProfile(auth.currentUser, {displayName: userName});
            // Setting display name to userEmail for now...
            await updateProfile(auth.currentUser, {displayName: userEmail});
            

            // add new user data to the firestore user list
            const newUser = {
                            email: userEmail,
                            favorites: []
                        }
            console.log(newUser);

            await setDoc(doc(db, "users", userEmail), newUser);

            setUserName('');
            setUserEmail('');
            setUserPassword('');
            setErrorMessage('');
            setDisplayName(userCredential.user.displayName);
        } catch(error) {
            console.log(error);
            showLoginError(error);
        }
    }

    const logout = async () => {
        await signOut(auth);
    }

    const showLoginError = (error) => {
        if (error.code === AuthErrorCodes.INVALID_PASSWORD) {
            setErrorMessage("Wrong password. Try again.");
        } else if (error.code === AuthErrorCodes.INVALID_EMAIL) {
            setErrorMessage("Email not recognized. Try again.");
        } else if (error.code === AuthErrorCodes.EMAIL_EXISTS) {
            setErrorMessage("Email already in use. Try again.");
        } else if (error.code === AuthErrorCodes.USER_DELETED) {
            setErrorMessage("User not found. Try again.");
        } else {
            setErrorMessage("Something went wrong: " + error.message);
        }
    }

    const handleChange = (event) => {
        if (event.target.id === "emailInput") {
            setUserEmail(event.target.value);
        } else if (event.target.id === "passwordInput") {
            setUserPassword(event.target.value);
        } else {
            setUserName(event.target.value);
        }
    }


    const [userName, setUserName] = useState('');
    const [userEmail, setUserEmail] = useState('');
    const [userPassword, setUserPassword] = useState('');
    const [errorMessage, setErrorMessage] = useState('');
    const [userSignedIn, setUserSignedIn] = useState(false);
    const [displayName, setDisplayName] = useState('');

    useEffect(() => {
        monitorAuthState()
    },[]);

    
    return (
        
        <div>
            {!userSignedIn &&
            <form className="loginForm">
                
                <label><h3>Sign in to view your favorites!!!</h3>
                    {/* <StyledInput
                        id="userNameInput"
                        placeholder="Enter your display name (optional)"
                        value={userName}
                        onChange={handleChange} /> */}
                    
                    <StyledInput
                        id="emailInput"
                        placeholder="Enter your email"
                        value={userEmail}
                        onChange={handleChange} />

                    <StyledInput
                        id="passwordInput"
                        placeholder="Enter your password"
                        value={userPassword}
                        onChange={handleChange} />
                </label>

                <div><i>{errorMessage}</i></div>
                <div>
                    <StyledButton
                        primary={true}
                        onClick={loginEmailPassword}
                        disabled={false} >
                            Sign In
                    </StyledButton>
                        - or -
                    <StyledButton
                        primary={false}
                        onClick={createAccount}
                        disabled={false}>
                            Sign Up
                    </StyledButton>
                </div>
            </form>}

            {userSignedIn &&
            <div>
            {/* <div>Welcome, {displayName}</div> */}
            <StyledButton
                        primary={false}
                        onClick={logout}
                        disabled={false}>
                            Log out
                    </StyledButton>
            </div>
            }
        
        
        </div>

    );
}

export default LoginPage;