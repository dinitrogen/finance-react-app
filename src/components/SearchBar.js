import React from "react";
import { StyledButton, StyledInput } from "./StyledComponents";


const SearchBar = ({clicked, ticker, handleChange, coolDown}) => {
    
    return (
        <div>
            <form className="searchForm">
                
                <label><h3>Ticker Search</h3>
                    <StyledInput
                        id="searchBar"
                        placeholder="Search for ticker (eg. VTI)"
                        value={ticker}
                        onChange={handleChange} />
                </label>
                <StyledButton
                    primary={!coolDown}
                    onClick={clicked}
                    disabled={coolDown}>
                        Search
                </StyledButton>
            </form>
            
        </div>
    );
}

export default SearchBar;