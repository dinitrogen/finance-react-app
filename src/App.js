import './App.css';
import React, { useEffect, useState } from "react";
import LandingPage from './components/LandingPage';
import SearchBar from './components/SearchBar';
import DisplayCard from './components/DisplayCard';
import { getEquityData } from './getEquityData';
import HistoryPage from './components/HistoryPage';
import uniqid from 'uniqid';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { collection, addDoc, setDoc, getDoc, doc, Timestamp, updateDoc, arrayUnion, deleteField, arrayRemove } from "firebase/firestore";
import { db, auth } from './firebase';
import FavoritesPage from './components/FavoritesPage';
import { StyledAppDiv } from './components/StyledComponents';
import LoginPage from './components/LoginPage';
import { onAuthStateChanged } from 'firebase/auth';

//let myFavorites = [];
async function checkForRepeats(newTickerName) {
  const docRef = doc(db, "stocks", "favorites");
  const docSnap = await getDoc(docRef);

  if (docSnap.exists()) {
    // console.log(docSnap.data().favorites);
    let currentFavorites = docSnap.data().favorites;
    //return docSnap.data().favorites;
    for (let favorite of currentFavorites) {
      // console.log(favorite["ticker"], newTickerName);
      if (favorite["ticker"] === newTickerName) {
        // console.log("Found repeat!");
        return false;
      }
    }
    // console.log("No repeats found");
    return true;

  } else {
    console.log("No such document!");
  }
}




const printNothing = () => {console.log("Nothing!")}

const App = () => {

  const [headerText, setHeaderText] = useState('');
  const [bodyText, setBodyText] = useState('');
  const [footerText, setFooterText] = useState('');
  const [ticker, setTicker] = useState('');
  const [tickers, setTickers] = useState([]);
  const [loading, setLoading] = useState(false);
  const [coolDown, setCoolDown] = useState(false);
  const [isFavorite, setIsFavorite] = useState(false);
  const [userName, setUserName] = useState('Guest');
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  
  useEffect(() => {
    monitorAuthState();
  },[]);

  const monitorAuthState = async () => {
    onAuthStateChanged(auth, user => {
        if (user) {
            setUserName(user.displayName);
            setIsLoggedIn(true);
            
        } else {
          setUserName('Guest');
          setIsLoggedIn(false); 
        }
    });
  }
  

  async function addToFavorites(tickerName) {
  
    // TODO: Check if ticker is already in favorites
    if (await checkForRepeats(tickerName)) {

      const newFavorite = {
        ticker: tickerName,
        date: Timestamp.fromDate(new Date()),
        id: uniqid()
      }
      
      // const firestoreFavorites = {
      //   favorites: myFavorites
      // }
    
      try {
        // updateDoc won't work unless a favorites array already exists
        await updateDoc(doc(db, "stocks", "favorites"), {favorites: arrayUnion(newFavorite)});
        console.log("added!");
        setIsFavorite(true);
        //console.log(myFavorites);
      } catch(error) {
        console.log("Error: ", error)
      }
    } else {
      console.log("Not added");
    }
    
  }

  const removeFromFavorites = async (tickerName) => {
    console.log("removed!");
    setIsFavorite(false);
    const docSnap = await getDoc(doc(db, "stocks", "favorites"));
    const arr = docSnap.data().favorites;
    // console.log(docSnap.data().favorites);
  
    const index = arr.findIndex(object => {
      return object.ticker === tickerName;
    });
    // console.log(index);

    await updateDoc(doc(db, "stocks", "favorites"), {favorites: arrayRemove(arr[index])});

  }

  const handleChange = (event) => {
      const value = event.target.value;
      setTicker(value);
  }

  const handleClick = (event) => {
    runEquitySearch(event);
    setTicker('');
  }

  const startCoolDownTimer = () => {
    console.log("start cooldown");
    setCoolDown(true);
    setTimeout(() => {
      setCoolDown(false);
      console.log("cooldown done!");
    }, 30000);
    
  }

  async function runEquitySearch(event) {
    event.preventDefault();
    setLoading(true);
    const searchResults = await getEquityData(ticker);
    setLoading(false);
    setHeaderText(searchResults[1]);
    setBodyText(searchResults[2]);
    setFooterText(searchResults[3]);
    if (searchResults[0]) {
      
      // TODO: add a button or toggle to add to favorites
      // addToFavorites(newTicker);
      if (await checkForRepeats(searchResults[1])) {
        setIsFavorite(false)
      } else setIsFavorite(true)
      
      let newTicker = {key: uniqid(),headerText: searchResults[1], bodyText: searchResults[2], footerText: searchResults[3]};
      setTickers([newTicker, ...tickers]);
    
    }
    if (searchResults[4]) {
      startCoolDownTimer();
    }
  }

  const clearSearchHistory = () => {
    setTickers([]);
  }

  return (
    <StyledAppDiv>
      <BrowserRouter>

      <div>
        <LandingPage />
        <br></br>
        <div>Welcome, {userName}</div>
        <Routes>
          {/* To get Firebase hosting to direct to /finance-react-app, I added a "redirect" to the firebase.json */}
          <Route path="/finance-react-app" element = {

            <div>
              <SearchBar
                ticker={ticker}
                clicked = {handleClick}
                handleChange = {handleChange}
                coolDown={coolDown} />

              {bodyText && !headerText &&
                <div>
                <div><i>{bodyText}</i></div>
                <div><i>{footerText}</i></div>
                </div>
              }
              
              {headerText && <DisplayCard
                loading={loading}
                headerText={headerText}
                bodyText={bodyText}
                footerText={footerText}
                showButton={(isLoggedIn) ? true: false}
                handlePrimary={!isFavorite}
                buttonText={(isFavorite) ? "Remove from favorites" : "Add to favorites"}
                handleClick={(!isFavorite) ? () => addToFavorites(headerText) : () => removeFromFavorites(headerText)}
                />}
            </div>} />
          
          <Route path="history" element = {
            <HistoryPage
              tickers={tickers}
              showButton={false}
              handleClick={clearSearchHistory} />} />
          <Route path="favorites" element = {
            <div>
              <LoginPage />
              <FavoritesPage
                handleClick={printNothing} />
                
            </div> } />
        </Routes>
        <br></br>
        <div><b>Note:</b> the API only allows 5 requests per minute, and 500 total requests per day.</div>
      </div>
      </BrowserRouter>
    </StyledAppDiv>
  );
}

export default App;
