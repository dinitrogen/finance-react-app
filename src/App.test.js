import React from 'react';
import { render, screen } from '@testing-library/react';
import "@testing-library/jest-dom";
import userEvent from '@testing-library/user-event';
import SearchBar from './components/SearchBar';
import LoginPage from './components/LoginPage';

import App from './App';

describe("Search Bar component", () => {
  it("renders correct heading", () => {
    const {getByRole} = render(<SearchBar />);
    expect(getByRole("heading").textContent).toMatch(/Ticker Search/i);
  });
});


describe("Login page", () => {
  it("renders login page", () =>{
    const { container } = render(<LoginPage />);
    expect(container).toMatchSnapshot();
  });
});

describe("Search bar input", () => {
  it("calls onChange correct number of times", () => {
    const onChangeMock = jest.fn();
    render(<SearchBar handleChange={onChangeMock} />);
    const input = screen.getByRole("textbox");

    userEvent.type(input, "VTI");

    expect(onChangeMock).toHaveBeenCalledTimes(3);
  });

  it("input has correct values", () => {
    const onChangeMock = jest.fn();
    render(<SearchBar handleChange={onChangeMock} />);
    const input = screen.getByRole("textbox");

    userEvent.type(input, "GOOG");

    expect(input).toHaveValue("GOOG");
  })

})