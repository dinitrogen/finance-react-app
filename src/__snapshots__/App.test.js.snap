// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders login page 1`] = `
<div>
  <div>
    <form
      class="loginForm"
    >
      <label>
        <h3>
          Sign in to view your favorites!!!
        </h3>
        <input
          class="sc-hKMtZM bXGjVj"
          id="emailInput"
          placeholder="Enter your email"
          value=""
        />
        <input
          class="sc-hKMtZM bXGjVj"
          id="passwordInput"
          placeholder="Enter your password"
          value=""
        />
      </label>
      <div>
        <i />
      </div>
      <div>
        <button
          class="sc-dkzDqf chaoPV"
        >
          Sign In
        </button>
        - or -
        <button
          class="sc-dkzDqf dYCMHp"
        >
          Sign Up
        </button>
      </div>
    </form>
  </div>
</div>
`;
